docs/spec_draft.md

```diff
@@ -102,26 +102,191 @@ never produced as outputs.
+
+```mlir
+stablehlo.func @main(
+  %image: tensor<28x28xf32>,
+  %weights: tensor<784x10xf32>,
+  %bias: tensor<1x10xf32>
+) -> tensor<1x10xf32> {
+  %0 = "stablehlo.reshape"(%image) : (tensor<28x28xf32>) -> tensor<1x784xf32>
+  %1 = "stablehlo.dot"(%0, %weights) : (tensor<1x784xf32>, tensor<784x10xf32>) -> tensor<1x10xf32>
+  %2 = "stablehlo.add"(%1, %bias) : (tensor<1x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
+  %3 = "stablehlo.constant"() { value = dense<0.0> : tensor<1x10xf32> } : () -> tensor<1x10xf32>
+  %4 = "stablehlo.maximum"(%2, %3) : (tensor<1x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
+  "stablehlo.return"(%4): (tensor<1x10xf32>) -> ()
 }
 ```
 
-A program is executed by passing argument values to a given function and
-computing output values. Output values of a function are computed by evaluating
-the graph of ops rooted in the corresponding return op. The evaluation order is
-implementation-defined, as long as ops are evaluated before their uses. Possible
-execution orders of the above example program are `%0` → `%1` → `%2` → `return`
-or `%1` → `%0` → `%2` → `return`.
+## Execution
+
+### Sequential execution
+
+A StableHLO program is executed by providing input values to the `main` function
+and computing output values. Output values of a function are computed by
+executing the graph of ops rooted in the corresponding `return` op.
+
+The execution order is implementation-defined, as long as ops are executed
+before their uses. Possible execution orders of the example program above are
+`%0` → `%1` → `%2` → `%3` → `%4` → `return` or `%3` → `%0` → `%1` → `%2` → `%4`
+→ `return`.
+
+More formally, a **StableHLO process** is a combination of:
+1) a StableHLO program, 2) operation statuses (not executed yet,
+already executed), and 3) intermediate values that the process is working on.
+The process starts with input values to the `main` function, progresses through
+the graph of ops updating operation statuses and intermediate values and
+finishes with output values. Further formalization is TBD.
+
+### Parallel execution
+
+StableHLO programs can be executed in parallel, organized into a 2D grid of
+`num_replicas` by `num_partitions` which both have type `ui32`.
+
+In the **StableHLO grid**, `num_replicas * num_partitions` of StableHLO
+processes are executing at the same time. Each process has a unique
+`process_id = (replica_id, partition_id)`, where
+`replica_id ∊ replica_ids = [0, ..., num_replicas-1]` and
+`partition_id ∊ partition_ids = [0, ..., num_partitions-1]` which both have
+type `ui32`. The size of the grid is known statically for every program, and
+the position within the grid is known statically for every process.
+
+Within the grid, the programs can all be the same (in the "Single Program,
+Multiple Data" style), can all be different (in the "Multiple Program, Multiple
+Data" style) or something in between. Each program has access to its position
+within the grid via the `replica_id` and `partition_id` ops.
```

Comment:
> Within the grid, the programs can all be the same...
> ...
> Each program has access to its position within the grid via the `replica_id` and `partition_id` ops.

The use of "program" is a bit confusing. Would this sbe clearer?
"
Within the grid, different processes can execute programs which are all the same...
...
Each *process* has access to its position....
"